components/coprocessor_plugin_api/src/allocator.rs

```diff
@@ -0,0 +1,53 @@
+// Copyright 2021 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::alloc::{GlobalAlloc, Layout};
+use std::sync::atomic::{AtomicPtr, Ordering};
+
+type AllocFn = unsafe fn(Layout) -> *mut u8;
+type DeallocFn = unsafe fn(*mut u8, Layout);
+
+/// Used to initialize the plugin's allocator.
+///
+/// A `HostAllocatorPtr` contains the relevant pointers to initialize the allocator of
+/// to plugin. It will be passed from TiKV to the plugin.
+#[repr(C)]
+pub struct HostAllocatorPtr {
+    pub alloc_fn: AllocFn,
+    pub dealloc_fn: DeallocFn,
+}
+
+/// An allocator that forwards invocations to the host (TiKV) of the plugin.
+pub struct HostAllocator {
+    alloc_fn: AtomicPtr<AllocFn>,
+    dealloc_fn: AtomicPtr<DeallocFn>,
+}
+
+impl HostAllocator {
+    /// Creates a new [`HostAllocator`], but initializes the functions to the current allocator.
+    /// Make sure to call [`set_allocator()`] because otherwise the default allocator will be used.
+    pub const fn new() -> Self {
+        HostAllocator {
+            alloc_fn: AtomicPtr::new(std::alloc::alloc as *mut _),
```

Comment:
I'll prefer using null point as default, so it'll segment fault immediately on startup when the host allocator is not set, instead of popping up some mystery bugs after dealloc.

Reply:
And in this case, when a plugin defined this as the global allocator, and for some reasons the allocator ptr is not set, any allocation will result in dead loop.

Reply:
yes, good point! Will change it :)